File: kotlin/Two Pointers/TripletSumBruteForce.kt
fun tripletSumBruteForce(nums: List<Int>): List<List<Int>> {
    // Use a hash set to ensure we don't add duplicate triplets.
    val triplets = hashSetOf<List<Int>>()
    // Iterate through the indexes of all triplets. O(n^3)
    for (i in nums.indices) {
        for (j in i + 1 until nums.size) {
            for (k in j + 1 until nums.size) {
                if (nums[i] + nums[j] + nums[k] == 0) {
                    // Sort the triplet before including it in the hash set.
                    val candidateList = mutableListOf(nums[i], nums[j], nums[k])
                    candidateList.sort()
                    triplets.add(candidateList)
                }
            }
        }
    }
    return triplets.toList()
}

File: typescript/Two Pointers/triplet_sum_brute_force.ts
function tripletSumBruteForce(nums: number[]): number[][] {
    let n = nums.length;
    // Use a hash set to ensure we don't add duplicate triplets.
    let triplets: Set<string> = new Set();
    // Iterate through the indexes of all triplets.
    for (let i = 0; i < n - 2; i++) {
        for (let j = i + 1; j < n - 1; j++) {
            for (let k = j + 1; k < n; k++) {
                if (nums[i] + nums[j] + nums[k] === 0) {
                    // Sort the triplet before including it in the hash set.
                    // [javascript] convert triplet array into a JSON string to use as a unqiue Set value.
                    let triplet = JSON.stringify([nums[i], nums[j], nums[k]].sort())
                    triplets.add(triplet)
                }
            }
        }
    }
    // [javascript] convert the Set back into an array of triplets.
    return Array.from(triplets).map((str) => JSON.parse(str));
}
